jupyter_sphinx/execute.py

```diff
@@ -0,0 +1,373 @@
+            except StopIteration:
+                continue
+
+            data = output['data'][mime_type]
+            if mime_type.startswith('image'):
+                # Sphinx treats absolute paths as being rooted at the source
+                # directory, so make a relative path, which Sphinx treats
+                # as being relative to the current working directory.
+                filename = os.path.basename(
+                    output.metadata['filenames'][mime_type]
+                )
+                uri = os.path.join(dir, filename)
+                to_add.append(nodes.image(uri=uri))
+            elif mime_type == 'text/html':
+                to_add.append(nodes.raw(
+                    text=data,
+                    format='html'
+                ))
+            elif mime_type == 'text/latex':
+                to_add.append(displaymath(
+                    latex=data,
+                    nowrap=False,
+                    number=None,
+                 ))
+            elif mime_type == 'text/plain':
+                to_add.append(nodes.literal_block(
+                    text=data,
+                    rawsource=data,
+                    language='ipython',
+                ))
+
+    return to_add
+
+
+def attach_outputs(output_nodes, node):
+    if node.attributes['hide_code']:
+        node.children = []
+    if not node.attributes['hide_output']:
+        if node.attributes['code_below']:
+            node.children = output_nodes + node.children
+        else:
+            node.children = node.children + output_nodes
+
+
+def default_notebook_names(basename):
+    """Return an interator yielding notebook names based off 'basename'"""
+    yield basename
+    for i in count(1):
+        yield '_'.join((basename, str(i)))
+
+
+def execute_cells(kernel_name, cells, execute_kwargs):
+    """Execute Jupyter cells in the specified kernel and return the notebook."""
+    notebook = blank_nb(kernel_name)
+    notebook.cells = cells
+    # Modifies 'notebook' in-place
+    try:
+        executenb(notebook, **execute_kwargs)
+    except Exception as e:
+        raise ExtensionError('Notebook execution failed', orig_exc=e)
+
+    return notebook
+
+
+def write_notebook_output(notebook, output_dir, notebook_name):
+    """Extract output from notebook cells and write to files in output_dir.
+
+    This also modifies 'notebook' in-place, adding metadata to each cell that
+    maps output mime-types to the filenames the output was saved under.
+    """
+    resources = dict(
+        unique_key=os.path.join(output_dir, notebook_name),
+        outputs={}
+    )
+
+    # Modifies 'resources' in-place
+    ExtractOutputPreprocessor().preprocess(notebook, resources)
+    # Write the cell outputs to files where we can (images and PDFs),
+    # as well as the notebook file.
+    FilesWriter(build_directory=output_dir).write(
+        nbformat.writes(notebook), resources,
+        os.path.join(output_dir, notebook_name + '.ipynb')
+    )
```

Comment:
Do we necessarily need to write a notebook file to disk?

Reply:
This is in case we want to let the user download it. I think this is quite useful for more involved tutorials.

Reply:
Gotcha, although it exposes the "notebook" nature of the directive.

Reply:
I addressed the naming of the directive options elsewhere, but I believe that we should keep writing the notebooks. Are we in agreement on that?